### DISCUSSION TITLE
[Help needed]  TS issues when creating a store factory.

### DISCUSSION THREAD

**Discussion by DonatelloTHM** [pmndrs/zustand#1234] [Q&A] [answered]
Hi, I'm trying to create a small helper to generate a store with `immer, subscribeWithSelector, devtools` middlewares. I'm not sure if I got the types right as I'm getting `error TS2349: This expression is not callable.
  Type 'never' has no call signatures.` everywhere I'm invoking `set` or `get`.

`Zustand 4.1.1`
```ts
export const createStoreWithSubscribe = <T>(
  config: StateCreator<T, [['zustand/subscribeWithSelector', T], ['zustand/immer', T], ['zustand/devtools', T]], []>,
  options: Parameters<typeof devtools>[1]
): StoreReturnType<T> => create<T>()(subscribeWithSelector(immer(devTools(config, options))))
```

I also tried to explicitly specify the return type:
```ts
type StoreReturnType<T> = UseBoundStore<
  Mutate<
    StoreApi<T>,
    [['zustand/subscribeWithSelector', never], ['zustand/immer', never], ['zustand/devtools', never], ...[]]
  >
>;

export const createStoreWithSubscribe = <T>(
  config: StateCreator<T, [['zustand/subscribeWithSelector', T], ['zustand/immer', T], ['zustand/devtools', T]], []>,
  options: Parameters<typeof devtools>[1]
): StoreReturnType<T> => create<T>()(subscribeWithSelector(immer(devTools(config, options)))) as StoreReturnType<T>;
```
That didn't really help.

Previously, prior to v4.0 this was my implementation that used a custom immer middleware, that worked fine.

```ts
export const createStoreWithPersistAndSubscribe = <T extends State>(
  config: StateCreator<
    T,
    (partial: ((draft: Draft<T>) => void) | T | Partial<T>, replace?: boolean, type?: string) => void,
    GetState<T>,
    Mutate<StoreApi<T>, [['zustand/subscribeWithSelector', never]]> &
      Mutate<StoreApi<T>, [['zustand/persist', Partial<T>]]> &
      Mutate<StoreApi<T>, [['zustand/devtools', never]]> &
      StoreApi<T>
  >,
  persistOptions: PersistOptions<T>,
  options: Parameters<typeof devtools>[1]
) => create(devTools(subscribeWithSelector(persist(withImmer(config), persistOptions)), options));
```

I'd appreciate anyone's help! Thanks.

**Comment by dai-shi** [comment]
Can you create a small reproduction with https://www.typescriptlang.org/play/ ?

also, check out docs, if you haven’t.

**Comment by DonatelloTHM** [comment, marked as answer]
Sorry, it seems like the issue is with our nx monorepo. For some reason when building, `immer` midleware types are not getting resolved, that's why the  type of the store instance is always `never`. I even copied the `immer` middleware from this repo and added a custom type like this:
```ts
declare module 'zustand' {
  interface StoreMutators<S, A> {
    'zimmer': WithDevtools<S>
  }
}
```
but it's not getting applied.
I guess we could close this discussion, as the issue doesn't seem to be related to `zustand`.
Thanks for your help!
